src/graph/LookupExecutor.cpp

```diff
@@ -0,0 +1,499 @@
+                                       expr->toString().c_str());
+        }
+        default : {
+            return Status::SyntaxError("Unsupported syntax ： %s", expr->toString().c_str());
+        }
+    }
+    return Status::OK();
+}
+
+Status LookupExecutor::checkFilter() {
+    auto status = traversalExpr(whereWrapper_->getFilter());
+    return status;
+}
+
+Status LookupExecutor::chooseIndex() {
+    auto ret = (isEdge_) ? ectx()->getMetaClient()->getEdgeIndexesFromCache(spaceId_)
+                         : ectx()->getMetaClient()->getTagIndexesFromCache(spaceId_);
+    if (!ret.ok()) {
+        return Status::IndexNotFound();
+    }
+
+    auto indexes = ret.value();
+    if (indexes.empty()) {
+        return Status::IndexNotFound();
+    }
+
+    auto validIndexes = findValidIndex(indexes);
+    if (validIndexes.empty()) {
+        LOG(ERROR) << "No matching index was found";
+        return Status::Error("No matching index was found");
+    }
+    return findOptimalIndex(validIndexes);
+}
+
+/**
+ * Choose the index that filter best matches, for example :
+ * the filter is col1 == 1 and col2 == 2 and col3 == 3,
+ * the indexes are index1 (col4, col3, col2, col1)
+ *                 index2 (col1, col2, col3, col4)
+ *                 index3 (col1, colA, col2, col3)
+ * as above, these indexes all are valid, but the index2 should be choose.
+ */
+Status LookupExecutor::findOptimalIndex(std::vector<IndexID> ids) {
+    Status status = Status::OK();
+    if (ids.size() == 1) {
+        index_ = ids[0];
+        return status;
+    }
+    /**
+     * Returns the first index if no optimization is required
+     */
+    if (skipOptimize_) {
+        index_ = ids[0];
+        return status;
+    }
+    auto *mc = ectx()->getMetaClient();
+
+    /**
+     * std::map<column_hint_count, IndexID>
+     */
+    std::map<int32_t, IndexID> indexHint;
+    for (auto& indexId : ids) {
+        auto itemStatus = (isEdge_) ? mc->getEdgeIndexFromCache(spaceId_, indexId)
+                                    : mc->getTagIndexFromCache(spaceId_, indexId);
+        if (!itemStatus.ok()) {
+            return itemStatus.status();
+        }
+        const auto& index = itemStatus.value();
+        int32_t hintCount = 0;
+        for (const auto& field : index->get_fields()) {
+            auto it = std::find_if(filters_.begin(), filters_.end(),
+                                   [field](const auto &rel) {
+                                       return field.get_name() == rel.first &&
+                                              rel.second == RelationalExpression::EQ;
+                                   });
+            if (it == filters_.end()) {
+                break;
+            }
+            ++hintCount;
+        }
+        /**
+         * return the index of max hint count.
+         */
+        indexHint[hintCount] = indexId;
+    }
+    index_ = indexHint.rbegin()->second;
```

Comment:
Why we can return the indexId with most hintCount? Like the example above: `index1 (col4, col3, col2, col1)` and `index2 (col1, col2, col3, col4)`, I suppose the hintCount of both index are 3?